src/rust/catnap/socket.rs

```diff
@@ -331,14 +181,235 @@ impl Socket {
+            )),
+            SocketOp::Listen => Err(fail(
+                op,
+                &(format!("socket is accepting connections on address: {:?}", self.local)),
+                libc::EADDRINUSE,
+            )),
+            SocketOp::Accept => Err(fail(
+                op,
+                &(format!("socket is accepting connections on address: {:?}", self.local)),
+                libc::EINPROGRESS,
+            )),
+            SocketOp::Accepted => Ok(SocketState::Listening),
+            SocketOp::Connect => Err(fail(
+                op,
+                &(format!("socket is accepting connections on address: {:?}", self.local)),
+                libc::ENOTSUP,
+            )),
+            // Should this be possible without going through the Connecting state?
+            SocketOp::Connected => Err(fail(
+                op,
+                &(format!("socket is accepting connections on address: {:?}", self.local)),
+                libc::EBUSY,
+            )),
+            SocketOp::Close => Ok(SocketState::Closing),
+            SocketOp::Closed => Err(fail(op, &(format!("socket is busy")), libc::EBUSY)),
+        }
+    }
+
+    fn connecting_state(&self, op: SocketOp) -> Result<SocketState, Fail> {
+        match op {
+            SocketOp::Bind => Err(fail(
+                op,
+                &(format!("socket is connecting to address: {:?}", self.remote)),
+                libc::EINVAL,
+            )),
+            SocketOp::Listen => Err(fail(
+                op,
+                &(format!("socket is connecting to address: {:?}", self.remote)),
+                libc::EADDRINUSE,
+            )),
+            SocketOp::Accept => Err(fail(
+                op,
+                &(format!("socket is connecting to address: {:?}", self.remote)),
+                libc::EINVAL,
+            )),
+            SocketOp::Accepted => Err(fail(
+                op,
+                &(format!("socket is connecting to address: {:?}", self.remote)),
+                libc::EINVAL,
+            )),
+            SocketOp::Connect => Err(fail(
+                op,
+                &(format!("socket already is connecting to address: {:?}", self.remote)),
+                libc::EINPROGRESS,
+            )),
+            SocketOp::Connected => Ok(SocketState::Connected),
+            SocketOp::Close => Ok(SocketState::Closing),
+            SocketOp::Closed => Err(fail(op, &(format!("socket is busy")), libc::EBUSY)),
+        }
+    }
+
+    fn connected_state(&self, op: SocketOp) -> Result<SocketState, Fail> {
+        match op {
+            // Does this make sense if we didn't go through the Connecting state?
+            SocketOp::Bind => Ok(SocketState::Connected),
+            SocketOp::Listen => Err(fail(
+                op,
+                &(format!("socket is already connected to address: {:?}", self.remote)),
+                libc::EISCONN,
+            )),
+            SocketOp::Accept => Err(fail(
```

Comment:
Many of these format strings and arguments are the same for several cases. It would be good to evaluate if these can be centralized. This comment applies to all the newly added functions.

Reply:
I collapsed them and committed in a separate commit. Please take a look.